File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/core/usecase/external/erp/usecases/ReturnsPromotion.java
package com.mycompany.myshop.testkit.dsl.core.usecase.external.erp.usecases;

import com.mycompany.myshop.testkit.driver.port.external.erp.ErpDriver;
import com.mycompany.myshop.testkit.driver.port.external.erp.dtos.ReturnsPromotionRequest;
import com.mycompany.myshop.testkit.dsl.core.usecase.external.erp.usecases.base.BaseErpUseCase;
import com.mycompany.myshop.testkit.common.Converter;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseResult;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseContext;
import com.mycompany.myshop.testkit.dsl.core.shared.VoidVerification;
import com.mycompany.myshop.testkit.driver.port.dtos.error.SystemError;

public class ReturnsPromotion extends BaseErpUseCase<Void, VoidVerification> {
    private boolean promotionActive;
    private String discount;

    public ReturnsPromotion(ErpDriver driver, UseCaseContext context) {
        super(driver, context);
    }

    public ReturnsPromotion withActive(boolean promotionActive) {
        this.promotionActive = promotionActive;
        return this;
    }

    public ReturnsPromotion withDiscount(String discount) {
        this.discount = discount;
        return this;
    }

    public ReturnsPromotion withDiscount(double discount) {
        return withDiscount(Converter.fromDouble(discount));
    }

    @Override
    public UseCaseResult<Void, VoidVerification> execute() {
        var request = ReturnsPromotionRequest.builder()
                .promotionActive(promotionActive)
                .discount(discount)
                .build();

        var result = driver.returnsPromotion(request);

        return new UseCaseResult<>(result.mapError(e -> SystemError.of(e.getMessage())), context, VoidVerification::new);
    }
}

File: system-test/typescript/src/testkit/dsl/core/usecase/external/erp/usecases/ReturnsPromotion.ts
import type { ErpDriver } from '../../../../../../driver/port/external/erp/erp-driver.js';
import type { ReturnsPromotionRequest } from '../../../../../../driver/port/external/erp/dtos/ReturnsPromotionRequest.js';
import { success, failure, type Result } from '../../../../../../common/result.js';
import type { SystemError } from '../../../../../../driver/port/dtos/errors/SystemError.js';
import { UseCaseResult } from '../../../../shared/use-case-result.js';
import { VoidVerification } from '../../../../shared/void-verification.js';
import type { UseCaseContext } from '../../../../shared/use-case-context.js';
import { BaseErpUseCase } from './base/BaseErpUseCase.js';

export class ReturnsPromotion extends BaseErpUseCase<void, VoidVerification> {
  private _promotionActive: boolean = false;
  private _discount: number | string = '';

  constructor(driver: ErpDriver, context: UseCaseContext) {
    super(driver, context);
  }

  withActive(promotionActive: boolean): this {
    this._promotionActive = promotionActive;
    return this;
  }

  withDiscount(discount: number | string): this {
    this._discount = discount;
    return this;
  }

  async execute(): Promise<UseCaseResult<void, VoidVerification>> {
    const request: ReturnsPromotionRequest = {
      promotionActive: this._promotionActive,
      discount: String(this._discount),
    };

    const driverResult = await this.driver.returnsPromotion(request);
    const result: Result<void, SystemError> = driverResult.success
      ? success(undefined)
      : failure({ message: driverResult.error.message ?? 'ERP error', fieldErrors: [] });

    return new UseCaseResult(
      result,
      this.context,
      (_, ctx) => new VoidVerification(undefined, ctx),
    );
  }
}
